org.eclipse.xtext.web/src/main/java/org/eclipse/xtext/web/server/model/UpdateDocumentService.java

```diff
@@ -0,0 +1,87 @@
+/**
+ * Copyright (c) 2015, 2020 itemis AG (http://www.itemis.eu) and others.
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ * 
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.eclipse.xtext.web.server.model;
+
+import org.eclipse.xtext.util.CancelIndicator;
+import org.eclipse.xtext.util.concurrent.CancelableUnitOfWork;
+import org.eclipse.xtext.web.server.InvalidRequestException;
+
+import com.google.inject.Singleton;
+
+/**
+ * Service class for document updates. This service only makes sense with a
+ * stateful server, where the server-side document representation is updated
+ * after each modification. This can greatly improve response times compared to
+ * the stateless alternative, where the full text content is sent with each
+ * service request.
+ */
+@Singleton
+public class UpdateDocumentService {
+	/**
+	 * Update the state identifier and return it. A background process is started
+	 * where the given text is assigned to the document and the services registered
+	 * in {@link PrecomputedServiceRegistry} are invoked.
+	 */
+	public DocumentStateResult updateFullText(XtextWebDocumentAccess document, String fullText)
+			throws InvalidRequestException {
+		return document.modify(new CancelableUnitOfWork<DocumentStateResult, IXtextWebDocument>() {
```

Comment:
Might want to use `tryModify` here to prevent possible NPE on line 36

Reply:
this is a web document which does not have tryModify

Reply:
Don't have the code at hand, does it need one?

Reply:
the web document has no tryModify

Reply:
i also guess we wont want to introduce new api here.
as the state is unclear.